src/topp/PercolatorAdapter.cpp

```diff
@@ -638,6 +649,219 @@ class PercolatorAdapter :
     }
     return EXECUTION_OK;
   }
+
+  ExitCodes readOSWFile_(std::string in_osw, std::string osw_level, std::stringstream& pin_output, double ipf_max_peakgroup_pep, double ipf_max_transition_isotope_overlap, double ipf_min_transition_sn) {
+
+    sqlite3 *db;
+    sqlite3_stmt * stmt;
+    int  rc;
+    std::string select_sql;
+
+    // Open database
+    rc = sqlite3_open(in_osw.c_str(), &db);
+    if ( rc )
+    {
+      fprintf(stderr, "Can't open database: %s\n", sqlite3_errmsg(db));
+    }
+
+    if (osw_level == "ms1") {
+      select_sql = "SELECT *, RUN_ID || '_' || PRECURSOR.ID AS GROUP_ID FROM FEATURE_MS1 INNER JOIN (SELECT ID, PRECURSOR_ID, RUN_ID FROM FEATURE) AS FEATURE ON FEATURE_ID = FEATURE.ID INNER JOIN (SELECT ID, DECOY FROM PRECURSOR) AS PRECURSOR ON FEATURE.PRECURSOR_ID = PRECURSOR.ID INNER JOIN PRECURSOR_PEPTIDE_MAPPING ON PRECURSOR.ID = PRECURSOR_PEPTIDE_MAPPING.PRECURSOR_ID INNER JOIN (SELECT ID, MODIFIED_SEQUENCE FROM PEPTIDE) AS PEPTIDE ON PRECURSOR_PEPTIDE_MAPPING.PEPTIDE_ID = PEPTIDE.ID;";
+    } else if (osw_level == "transition") {
+      select_sql = "SELECT TRANSITION.DECOY AS DECOY, FEATURE_TRANSITION.*, RUN_ID || '_' || FEATURE_TRANSITION.FEATURE_ID || '_' || PRECURSOR_ID || '_' || TRANSITION_ID AS GROUP_ID, FEATURE_TRANSITION.FEATURE_ID || '_' || FEATURE_TRANSITION.TRANSITION_ID AS FEATURE_ID, 'PEPTIDE' AS MODIFIED_SEQUENCE FROM FEATURE_TRANSITION INNER JOIN (SELECT RUN_ID, ID, PRECURSOR_ID FROM FEATURE) AS FEATURE ON FEATURE_TRANSITION.FEATURE_ID = FEATURE.ID INNER JOIN PRECURSOR ON FEATURE.PRECURSOR_ID = PRECURSOR.ID INNER JOIN SCORE_MS2 ON FEATURE.ID = SCORE_MS2.FEATURE_ID INNER JOIN (SELECT ID, DECOY FROM TRANSITION) AS TRANSITION ON FEATURE_TRANSITION.TRANSITION_ID = TRANSITION.ID WHERE PEP <= " + String(ipf_max_peakgroup_pep) + " AND VAR_ISOTOPE_OVERLAP_SCORE <= " + String(ipf_max_transition_isotope_overlap) + " AND VAR_LOG_SN_SCORE > " + String(ipf_min_transition_sn) + " AND PRECURSOR.DECOY == 0 ORDER BY FEATURE_ID, PRECURSOR_ID, TRANSITION_ID;";
+    } else {
+      // Peak group-level query including peptide sequence
+      select_sql = "SELECT *, RUN_ID || '_' || PRECURSOR.ID AS GROUP_ID FROM FEATURE_MS2 INNER JOIN (SELECT ID, PRECURSOR_ID, RUN_ID FROM FEATURE) AS FEATURE ON FEATURE_ID = FEATURE.ID INNER JOIN (SELECT ID, DECOY FROM PRECURSOR) AS PRECURSOR ON FEATURE.PRECURSOR_ID = PRECURSOR.ID INNER JOIN PRECURSOR_PEPTIDE_MAPPING ON PRECURSOR.ID = PRECURSOR_PEPTIDE_MAPPING.PRECURSOR_ID INNER JOIN (SELECT ID, MODIFIED_SEQUENCE FROM PEPTIDE) AS PEPTIDE ON PRECURSOR_PEPTIDE_MAPPING.PEPTIDE_ID = PEPTIDE.ID;";
+    }
+
+    // Execute SQL select statement
+    sqlite3_prepare(db, select_sql.c_str(), -1, &stmt, NULL);
+    sqlite3_step( stmt );
+
+    int cols = sqlite3_column_count(stmt);
+
+    // Generate features
+    int k = 0;
+    std::vector<std::string> group_id_index;
+    std::vector<std::string> features_names;
+
+    while (sqlite3_column_type( stmt, 0 ) != SQLITE_NULL)
```

Comment:
I wonder whether we could put the sqlite specific stuff into the main library somewhere so that we dont have the abstractions leaking so much